Add "Connect Stripe Account" UI

Mayisha opened the pull request:
Fixes #1757

### Description
If  `_wcstripe_feature_upe_settings` is enabled and a Stripe account is not connected a `card` containing the related information and `Connect` button appears in the main Stripe settings page as well as in other non-UPE legacy payment method pages.

### Testing instructions

- Disconnect Stripe account (clear all the keys from old UI)
- Enable the `_wcstripe_feature_upe_settings` flag
- Check the main Stripe settings page and other legacy payment method settings pages

Desktop view of the Stripe settings page
![main stripe page](https://user-images.githubusercontent.com/33387139/131618036-0faf89b9-0f26-448a-a178-8c4718eee5a6.png)

Mobile view of the Stripe settings page
![mobile view stripe settings](https://user-images.githubusercontent.com/33387139/131618061-0f014184-27fa-4644-b875-6ca91938fd02.png)

Desktop view of the other methods' settings page
![legacy page](https://user-images.githubusercontent.com/33387139/131618084-f3d79ada-f413-40e3-8b67-b54f867122c0.png)

### QA checklists

- [ ] If  `_wcstripe_feature_upe_settings`  is disabled the old settings page is visible in Stripe and other payment method settings pages
- [ ] If  `_wcstripe_feature_upe_settings`  is enabled and Stripe is not connected, the `Connect Stripe Account` card is visible in Stripe settings page
- [ ]  If  `_wcstripe_feature_upe_settings`  is enabled and Stripe is not connected, the `Connect Stripe Account` card is visible in  all legacy method settings pages
- [ ]  If  `_wcstripe_feature_upe_settings`  is enabled and Stripe is connected, the `Connect Stripe Account` card should not be present

Mayisha committed:
Add Stripe banner as a component

Mayisha committed:
Add 'ConnectStripeAccount' component

Mayisha committed:
Add tests for 'ConnectStripeAccount' component

Mayisha committed:
Render 'ConnectStripeAccount' card in Stripe settings page
- Add placeholder div for the card when Stripe is connected
- Render card in ReactDom within the placeholder div
- Add styles to align the card at center of the page
- Fix stylesheet file name

Mayisha committed:
Render 'ConnectStripeAccount' card in other payment method's settings page when Stripe is not connected

waclawjacek commented on client/settings/connect-stripe-account/index.js:
```diff
@@ -0,0 +1,107 @@
+	<CardWrapper>
+		<StripeBanner />
+		<CardBody>
+			<h2>Get started with Stripe</h2>
```
These should be translatable:

```suggestion
			<h2>{ __( 'Get started with Stripe', 'woocommerce-gateway-stripe' ) }</h2>
```

waclawjacek commented on client/settings/connect-stripe-account/index.js:
```diff
@@ -0,0 +1,107 @@
+				Connect or create a Stripe account to accept payments directly
+				onsite, including Payment Request buttons (such as Apple Pay and
+				Google Pay), iDeal, SEPA, Sofort, and more international payment
+				methods.
```
```suggestion
				{ __( 'Connect or create a Stripe account to accept payments directly
				onsite, including Payment Request buttons (such as Apple Pay and
				Google Pay), iDeal, SEPA, Sofort, and more international payment
				methods.', 'woocommerce-gateway-stripe' ) }
```

waclawjacek commented on client/settings/connect-stripe-account/index.js:
```diff
@@ -0,0 +1,107 @@
+	color: #1e1e1e;
+`;
+
+const TermOfServicesText = styled.p`
```
This should probably be:

```suggestion
const TermsOfServiceText = styled.p`
```

waclawjacek commented on client/settings/styles.scss:
```diff
@@ -1,3 +1,12 @@
+
+	@media ( max-width: 660px ) {
+		display: block;
+	}
```
I think this could be removed. There's no harm in having it be `display: flex` all the time. :) With this left in place there is a very small range of widths where the card will be aligned left with some space to the right.

```suggestion
```

Mayisha replied:
Right. I was supposed to remove this but forgot :smile:

waclawjacek commented on client/settings/connect-stripe-account/index.js:
```diff
@@ -0,0 +1,107 @@
+				methods.
+			</InformationText>
+			<TermOfServicesText>
+				By clicking &quot;Create or connect an account&quot;, you agree
+				to{ ' ' }
+				<a
+					href="https://stripe.com/ssa"
+					target="_blank"
+					rel="noreferrer"
+				>
+					Stripe’s Terms of service.
+				</a>
```
You could use `interpolate-components` to i18n this. [Example here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/01f360c81df33b3ced998e2d7fd7937a4d86aacf/client/settings/payment-request-section/index.js#L73).

waclawjacek commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -23,11 +23,14 @@ public function __construct() {
 		echo '</h2>';
-		echo '<div id="wc-stripe-account-settings-container"></div>';
+
+		echo $is_stripe_connected ? '<div id="wc-stripe-account-settings-container"></div>' : '<div id="wc-stripe-new-account-container"></div>';
```
It would be great if you could add a test checking if the expected output is returned based on `woocommerce_gateway_stripe()->connect->is_connected()`.

waclawjacek commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -34,17 +34,24 @@ public function display_admin_settings_webhook_description() {
-			echo '<div id="wc-stripe-upe-opt-in-banner"></div>';
-			echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, $target_index, null, true ), false ) . '</table>';
+			if ( ! $is_stripe_connected ) {
+				global $hide_save_button;
```
This might just be personal preference (I didn't find any references about this after a quick search) but I think moving `global` to the top of the method body is a good idea - it makes sure you're using the same variable throughout the entire function and I think I've seen globals usually defined at the top which is where I'd usually look for them.

waclawjacek commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -34,17 +34,24 @@ public function display_admin_settings_webhook_description() {
-			echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, 0, $target_index, true ), false ) . '</table>';
-			echo '<div id="wc-stripe-upe-opt-in-banner"></div>';
-			echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, $target_index, null, true ), false ) . '</table>';
+			if ( ! $is_stripe_connected ) {
+				global $hide_save_button;
+				$hide_save_button  = true;
+				echo '<div id="wc-stripe-new-account-container"></div>';
+			} else {
+				echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, 0, $target_index, true ), false ) . '</table>';
+				echo '<div id="wc-stripe-upe-opt-in-banner"></div>';
+				echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, $target_index, null, true ), false ) . '</table>';
+			}
```
How about extracting this nested condition to a separate method to avoid nesting? E.g. `$this->render_upe_settings()`.

waclawjacek commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -34,17 +34,24 @@ public function display_admin_settings_webhook_description() {
-			echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, 0, $target_index, true ), false ) . '</table>';
-			echo '<div id="wc-stripe-upe-opt-in-banner"></div>';
-			echo '<table class="form-table">' . $this->generate_settings_html( array_slice( $form_fields, $target_index, null, true ), false ) . '</table>';
+			if ( ! $is_stripe_connected ) {
```
It would be great if you could add a test if the expected `<div id="">` is printed if possible.

waclawjacek commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -59,7 +62,7 @@ public function admin_scripts() {
 			wp_register_style(
 				'woocommerce_stripe_admin',
-				plugins_url( 'build/style-upe_settings.css', WC_STRIPE_MAIN_FILE ),
+				plugins_url( 'build/upe_settings.css', WC_STRIPE_MAIN_FILE ),
```
Good catch! :tada:

waclawjacek commented on client/settings/connect-stripe-account/index.js:
```diff
@@ -0,0 +1,107 @@
+		&:active,
+		&:focus,
+		&:hover {
+			box-shadow: none !important;
+			background: none !important;
```
Is it possible to avoid using `!important` here and use selectors with a higher specificity instead?

waclawjacek replied:
Your solution to this is good and no change needed here but I am wondering if we should maybe go with something like this instead (not sure if it's possible with `@emotion/styled` but it is with SCSS):

```
.button-wrapper {
  &#{&} {
    &#{&} {
      /* css here */
    }
  }
}
```

Essentially this results in this:

```
.button-wrapper.button-wrapper.button-wrapper { }
```

waclawjacek reviewed (changes requested):
Great work! :tada: Tested and works great!

I'm clicking "Request changes" but only because of the i18n stuff and the `!important` selector which I'm hoping is avoidable. The other comments are nitpicks.

I have noticed two things which I think we should take into account:

## Toggling live/test mode vs. key presence

Let's say you are using test mode. In the current UI, when you do a fresh installation, enable test mode and input the test API keys, the live keys are blank but everything will works as expected. This is good.

However, when you toggle the checkbox so it says "live mode" now and you don't input live keys, you'll get the Connect Stripe Account card displayed now, even though you have some keys that are working, you just don't have the ones for the currently-selected mode (live/test).

The same probably happens if you have live keys but you switch to test mode and don't input test keys - you'll get the Connect Stripe Account card.

I am wondering if this could be misleading to merchants and if there is something we could do about it. Maybe this is a question for a separate issue.

## Payment methods page notice

On payment method pages, you'll also get this notice:

![Markup on 2021-09-01 at 11:09:23](https://user-images.githubusercontent.com/1759681/131657772-58edc83e-afcc-4b2a-ab21-44d6836bcb8c.png)

Now that the Connect Account card is displayed on the payment method pages anyway, we should probably remove this notice. Clicking the link in it will just get you to the main gateway page which displays the same card anyway.

Mayisha commented on client/settings/connect-stripe-account/index.js:
```diff
@@ -0,0 +1,107 @@
+				By clicking &quot;Create or connect an account&quot;, you agree
+				to{ ' ' }
+				<a
+					href="https://stripe.com/ssa"
```
@waclawjacek could you please verify if I have added the correct link here?

waclawjacek replied:
Looks good to me! @jarekmorawski could you please verify if this is the correct page?

jarekmorawski replied:
I think so! Currently in the plugin we link to [WP's ToS](https://wordpress.com/tos/), but I'm not sure if that's the case here. @LevinMedia, do you have any idea if we should stick to WP's ToS?

waclawjacek replied:
I have posted a separate issue for this (#1828) so we can close this PR.

Mayisha committed:
Refactor code
-  i18n strings
- fix const name
- use 'interpolateComponents'
- Extract out rendering of UPE settings in a separate method

Mayisha committed:
Add test for asserting placeholder divs

Mayisha committed:
Remove and update unnecessary styles

Mayisha commented:
> Let's say you are using test mode. In the current UI, when you do a fresh installation, enable test mode and input the test API keys, the live keys are blank but everything will works as expected. This is good.
> 
> However, when you toggle the checkbox so it says "live mode" now and you don't input live keys, you'll get the Connect Stripe Account card displayed now, even though you have some keys that are working, you just don't have the ones for the currently-selected mode (live/test).
> 
> The same probably happens if you have live keys but you switch to test mode and don't input test keys - you'll get the Connect Stripe Account card.
> 
> I am wondering if this could be misleading to merchants and if there is something we could do about it. Maybe this is a question for a separate issue.

In the above scenario [this function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/9aac6a03bb7b5fcf8a93e55c22c22f97ff44c7f4/includes/connect/class-wc-stripe-connect.php#L185) returns `false` and indicates that Stripe is not connected. If we follow this logic then showing the `Connect Stripe Account` card seems like the expected behavior (as for the selected mode we don't have the keys). Even in the old UI, the key fields remain empty in this scenario. :thinking: 

Do you have any other ideas for this? Like if the test key is present when we are in live mode, let the merchant know about this somehow? Maybe we would need a design for this if we want to show that. I also think this can be handled as a separate issue.

waclawjacek commented:
> Do you have any other ideas for this? Like if the test key is present when we are in live mode, let the merchant know about this somehow? Maybe we would need a design for this if we want to show that. I also think this can be handled as a separate issue.

Let's leave this as it is in this PR and tackle this separately as we discussed during the call. :+1:

Mayisha committed:
Hide notice from payment method settings pages when upe redesign flag is enabled

waclawjacek commented on includes/admin/class-wc-stripe-admin-notices.php:
```diff
@@ -174,8 +174,10 @@ public function stripe_check_environment() {
 			if ( empty( $show_keys_notice ) ) {
 				$secret = WC_Stripe_API::get_secret_key();
+				// if '_wcstripe_feature_upe_settings' is enabled show the notice in main settings page only
+				$should_show_notice_on_page = WC_Stripe_Feature_Flags::is_upe_settings_redesign_enabled() ? ( isset( $_GET['page'] ) && ! isset( $_GET['section'] ) ) : ! ( isset( $_GET['page'], $_GET['section'] ) && 'wc-settings' === $_GET['page'] && 'stripe' === $_GET['section'] );
```
Thank you for including this change and great idea to extract this part of the condition into a variable!

Once thing that I noticed is that since the code is not checking for the value of "page" and "section" in the new part of the condition:

```php
( isset( $_GET['page'] ) && ! isset( $_GET['section'] ) )
```

This actually causes the condition **not** to appear where it used to appear. So e.g.:

* It will still appear on http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=shipping (this is good).
* It will no longer appear on http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=shipping&section=options (unlike previously).

So I think the way to solve this is to repeat these checks for when the UPE settings redesign is enabled:

```php
'wc-settings' === $_GET['page'] && 'stripe' === $_GET['section']
```

But also add a condition checking if the currently-visited page is a matching gateway page. All gateway names start with "stripe_" so this should work:

```php
0 === substr( $_GET['section'], 'stripe_' )
```

Mayisha replied:
Good catch. I will fix this :+1:

Mayisha replied:
Fixed the bug in [this commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1774/commits/c514af44097b6a716736e872dc3907ec8c8e09c5). 

Instead of matching the `stripe_` substring, matched the `stripe` substring so that all the other gateways as well as the stripe itself is covered in the logic.

waclawjacek replied:
Good call! It doesn't make sense to display it for the main gateway too.

Sorry to be so picky and this is technically correct but it is a pretty long condition that requires some grokking. And my bad for mentioning `substr()` - I actually had `strpos()` in mind!

How about this instead? Let me know if I missed something that your idea covers that this one doesn't!

```php
$is_stripe_payment_method_page = isset( $_GET['page'], $_GET['section'] ) && 'wc-settings' === $_GET['page'] && 0 === strpos( $_GET['section'], 'stripe' );
$should_show_notice_on_page    = WC_Stripe_Feature_Flags::is_upe_settings_redesign_enabled() && ! $is_stripe_payment_method_page;
```

waclawjacek replied:
I'm looking at this again condition again and I get the idea now - if the feature flag then hide the notice from the main Stripe gateway and the payment method pages, and if the flag is disabled - just the Stripe gateway? Makes sense! :)

I don't think the other filters are necessary though - there's no harm if someone tries to pass a malicious value since the code is only checking if it equals something (and not, e.g., displaying it or feeding it into a DB) so I'd just go with:

```php
$is_stripe_payment_method_page = isset( $_GET['page'], $_GET['section'] ) && 'wc-settings' === $_GET['page'] && 0 === strpos( $_GET['section'], 'stripe' );
$is_stripe_main_gateway_page   = isset( $_GET['page'], $_GET['section'] ) && 'wc-settings' === $_GET['page'] && 'stripe' === $_GET['section'];
$should_show_notice_on_page    = WC_Stripe_Feature_Flags::is_upe_settings_redesign_enabled() ? ! $is_stripe_payment_method_page : ! $is_stripe_main_gateway_page;
```

You can probably even simplify it to not distinguish the cases when UPE is enabled and just doing:

```php
$should_show_notice_on_page = ! ( isset( $_GET['page'], $_GET['section'] ) && 'wc-settings' === $_GET['page'] && 0 === strpos( $_GET['section'], 'stripe' ) );
```

waclawjacek commented on tests/phpunit/admin/test-class-wc-stripe-settings-controller.php:
```diff
@@ -0,0 +1,69 @@
+		ob_start();
+		$this->controller->admin_options( $this->gateway );
+		$output = ob_get_clean();
+		$this->assertStringMatchesFormat( '%aid="wc-stripe-account-settings-container"%a', $output );
```
Awesome! :rocket:

waclawjacek commented on tests/phpunit/admin/test-class-wc-stripe-settings-controller.php:
```diff
@@ -0,0 +1,69 @@
+
+	public function tearDown() {
+		parent::tearDown();
+		delete_option( 'woocommerce_stripe_settings' );
```
No change required here but I think this shouldn't be necessary because a fresh environment is set up for every test class anyway.

waclawjacek reviewed (changes requested):
Thank you for including the changes! Great work and the code is still awesome.

Just one bug that I found in regard to when the notice is displayed. Could you please fix that? Other than that, it's ready to go!

waclawjacek commented on tests/phpunit/test-wc-stripe-payment-gateway-legacy-methods.php:
```diff
@@ -0,0 +1,66 @@
+ */
+
+
+class WC_Stripe_Payment_Gateway_Legacy_Methods_Test extends WP_UnitTestCase {
```
To keep the test names consistent with the class names which they are testing, I think it'd be a good idea to move the tests in this class to the existing `WC_Stripe_Payment_Gateway_Test` class.

Mayisha replied:
Moved these 2 tests inside the `WC_Stripe_Payment_Gateway_Test` class in [this commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1774/commits/9c2c1c2829b5d8123a3902e1a9a1218d461d8d78). But we need to initiate another gateway in the class now (for any of the legacy methods), as the existing gateway is for `Stripe` which overrides the `admin_options` function and can not be used to assert the desired divs.

waclawjacek replied:
Great catch and thank you for including this change!

Mayisha committed:
Fix condition to show key notice

Mayisha committed:
Move tests inside WC_Stripe_Payment_Gateway_Test class

waclawjacek reviewed (dismissed):
If you could please just address [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1774#discussion_r708962355), that'll be all!

harriswong commented:
Merging `develop` in here. I would like some code in latest develop and my branch for [this issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1759) is basing off from this.

Mayisha committed:
Address feedback

Mayisha commented:
> If you could please just address [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1774#discussion_r708962355), that'll be all!

@waclawjacek addressed that :man_dancing:

frosso reviewed (approved):
It looks like all review comments have been addressed, approving since Waclaw is AFK

frosso committed:
Merge branch 'develop' into add/1757-new-account-onboarding-ui
